File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/socialmedia/SocialMediaHomeFragment.java
package com.github.dedis.popstellar.ui.socialmedia;

import android.os.Bundle;
import android.util.Log;
import android.view.*;

import androidx.annotation.*;
import androidx.fragment.app.Fragment;
import androidx.fragment.app.FragmentManager;

import com.github.dedis.popstellar.R;
import com.github.dedis.popstellar.databinding.SocialMediaHomeFragmentBinding;
import com.github.dedis.popstellar.ui.lao.LaoActivity;
import com.github.dedis.popstellar.ui.lao.LaoViewModel;
import com.github.dedis.popstellar.utility.ActivityUtils;

import java.util.function.Supplier;

import dagger.hilt.android.AndroidEntryPoint;

/**
 * The purpose of this fragment is to provide a bottom nav bar and fragment container to social
 * media fragments. In itself this fragment does not provide any social media feature
 */
@AndroidEntryPoint
public class SocialMediaHomeFragment extends Fragment {

  private SocialMediaViewModel socialMediaViewModel;
  private SocialMediaHomeFragmentBinding binding;

  public static final String TAG = SocialMediaHomeFragment.class.getSimpleName();

  @Nullable
  @Override
  public View onCreateView(
      @NonNull LayoutInflater inflater,
      @Nullable ViewGroup container,
      @Nullable Bundle savedInstanceState) {
    binding = SocialMediaHomeFragmentBinding.inflate(inflater, container, false);

    LaoViewModel viewModel = LaoActivity.obtainViewModel(requireActivity());
    socialMediaViewModel =
        LaoActivity.obtainSocialMediaViewModel(requireActivity(), viewModel.getLaoId());

    setupBottomNavBar();
    openChirpList();
    return binding.getRoot();
  }

  private void setupBottomNavBar() {
    socialMediaViewModel
        .getBottomNavigationTab()
        .observe(
            getViewLifecycleOwner(),
            tab -> binding.socialMediaNavBar.setSelectedItemId(tab.getMenuId()));

    binding.socialMediaNavBar.setOnItemSelectedListener(
        item -> {
          SocialMediaTab tab = SocialMediaTab.findByMenu(item.getItemId());
          Log.i(TAG, "Opening tab : " + tab.getName());
          openBottomTab(tab);
          return true;
        });

    binding.socialMediaNavBar.setOnItemReselectedListener(item -> {});
    socialMediaViewModel.setBottomNavigationTab(SocialMediaTab.HOME);
  }

  private void openBottomTab(SocialMediaTab tab) {
    switch (tab) {
      case HOME:
        openChirpList();
        break;
      case SEARCH:
        openSearchTab();
        break;
      case FOLLOWING:
        openFollowingTab();
        break;
      case PROFILE:
        openProfileTab();
        break;
    }
  }

  private void openChirpList() {
    setCurrentFragment(
        getParentFragmentManager(), R.id.fragment_chirp_list, ChirpListFragment::newInstance);
  }

  private void openSearchTab() {
    setCurrentFragment(
        getParentFragmentManager(),
        R.id.fragment_social_media_search,
        SocialMediaSearchFragment::newInstance);
  }

  private void openFollowingTab() {
    setCurrentFragment(
        getParentFragmentManager(),
        R.id.fragment_social_media_following,
        SocialMediaFollowingFragment::newInstance);
  }

  private void openProfileTab() {
    setCurrentFragment(
        getParentFragmentManager(),
        R.id.fragment_social_media_profile,
        SocialMediaProfileFragment::newInstance);
  }

  /**
   * Set the current fragment in the container of the home fragment
   *
   * @param id of the fragment
   * @param fragmentSupplier provides the fragment if it is missing
   */
  public static void setCurrentFragment(
      FragmentManager manager, @IdRes int id, Supplier<Fragment> fragmentSupplier) {
    ActivityUtils.setFragmentInContainer(
        manager, R.id.fragment_container_social_media, id, fragmentSupplier);
  }
}

File: fe2-android/app/src/main/java/com/github/dedis/popstellar/ui/lao/socialmedia/SocialMediaHomeFragment.kt
package com.github.dedis.popstellar.ui.lao.socialmedia

import android.os.Bundle
import android.view.LayoutInflater
import android.view.MenuItem
import android.view.View
import android.view.ViewGroup
import androidx.annotation.IdRes
import androidx.fragment.app.Fragment
import androidx.fragment.app.FragmentManager
import com.github.dedis.popstellar.R
import com.github.dedis.popstellar.databinding.SocialMediaHomeFragmentBinding
import com.github.dedis.popstellar.ui.lao.LaoActivity
import com.github.dedis.popstellar.ui.lao.LaoActivity.Companion.addBackNavigationCallbackToEvents
import com.github.dedis.popstellar.ui.lao.LaoActivity.Companion.obtainSocialMediaViewModel
import com.github.dedis.popstellar.ui.lao.LaoActivity.Companion.obtainViewModel
import com.github.dedis.popstellar.ui.lao.socialmedia.SocialMediaTab.Companion.findByMenu
import com.github.dedis.popstellar.utility.ActivityUtils.setFragmentInContainer
import dagger.hilt.android.AndroidEntryPoint
import java.util.function.Supplier
import timber.log.Timber

/**
 * The purpose of this fragment is to provide a bottom nav bar and fragment container to social
 * media fragments. In itself this fragment does not provide any social media feature
 */
@AndroidEntryPoint
class SocialMediaHomeFragment : Fragment() {

  private lateinit var socialMediaViewModel: SocialMediaViewModel
  private lateinit var binding: SocialMediaHomeFragmentBinding

  override fun onCreateView(
      inflater: LayoutInflater,
      container: ViewGroup?,
      savedInstanceState: Bundle?
  ): View {
    binding = SocialMediaHomeFragmentBinding.inflate(inflater, container, false)

    val viewModel = obtainViewModel(requireActivity())
    socialMediaViewModel = obtainSocialMediaViewModel(requireActivity(), viewModel.laoId!!)
    socialMediaViewModel.checkValidPoPToken()

    setupBottomNavBar()
    openChirpList()
    handleBackNav()

    return binding.root
  }

  private fun setupBottomNavBar() {
    socialMediaViewModel.bottomNavigationTab.observe(viewLifecycleOwner) { tab: SocialMediaTab ->
      binding.socialMediaNavBar.selectedItemId = tab.menuId
    }

    binding.socialMediaNavBar.setOnItemSelectedListener { item: MenuItem ->
      val tab = findByMenu(item.itemId)
      Timber.tag(TAG).i("Opening tab : %s", tab.name)
      openBottomTab(tab)
      true
    }

    binding.socialMediaNavBar.setOnItemReselectedListener {}
    socialMediaViewModel.setBottomNavigationTab(SocialMediaTab.HOME)
  }

  private fun openBottomTab(tab: SocialMediaTab) {
    when (tab) {
      SocialMediaTab.HOME -> openChirpList()
      SocialMediaTab.SEARCH -> openSearchTab()
      SocialMediaTab.FOLLOWING -> openFollowingTab()
      SocialMediaTab.PROFILE -> openProfileTab()
    }
  }

  private fun openChirpList() {
    setCurrentFragment(parentFragmentManager, R.id.fragment_chirp_list) {
      ChirpListFragment.newInstance()
    }
  }

  private fun openSearchTab() {
    setCurrentFragment(parentFragmentManager, R.id.fragment_social_media_search) {
      SocialMediaSearchFragment.newInstance()
    }
  }

  private fun openFollowingTab() {
    setCurrentFragment(parentFragmentManager, R.id.fragment_social_media_following) {
      SocialMediaFollowingFragment.newInstance()
    }
  }

  private fun openProfileTab() {
    setCurrentFragment(parentFragmentManager, R.id.fragment_social_media_profile) {
      SocialMediaProfileFragment.newInstance()
    }
  }

  private fun handleBackNav() {
    addBackNavigationCallbackToEvents(requireActivity(), viewLifecycleOwner, TAG)
  }

  companion object {
    val TAG: String = SocialMediaHomeFragment::class.java.simpleName

    fun newInstance(): SocialMediaHomeFragment {
      return SocialMediaHomeFragment()
    }

    /**
     * Set the current fragment in the container of the home fragment
     *
     * @param id of the fragment
     * @param fragmentSupplier provides the fragment if it is missing
     */
    fun setCurrentFragment(
        manager: FragmentManager,
        @IdRes id: Int,
        fragmentSupplier: Supplier<Fragment>
    ) {
      setFragmentInContainer(manager, R.id.fragment_container_social_media, id, fragmentSupplier)
    }

    fun openFragment(manager: FragmentManager) {
      LaoActivity.setCurrentFragment(manager, R.id.fragment_social_media_home) {
        SocialMediaHomeFragment()
      }
    }
  }
}
